<a href="https://colab.research.google.com/github/aparna-2001/matrix-properties-calculator/blob/main/mini_project_matrix2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import csv
import zipfile
import io


zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

try:
    with zipfile.ZipFile(zip_file_path, 'r') as zf:
        print(f"Files found inside '{zip_file_path}': {zf.namelist()}\n")


        if csv_filename_in_zip in zf.namelist():
            with zf.open(csv_filename_in_zip) as csv_file:
                # zf.open returns a file-like object in binary mode by default.
                # csv.reader expects a text file, so we wrap it in io.TextIOWrapper
                text_stream = io.TextIOWrapper(csv_file, encoding='utf-8') # Or the correct encoding of your CSV
                csv_rows = list(csv.reader(text_stream))
                print(f"Successfully read {len(csv_rows)} rows from '{csv_filename_in_zip}'.")

                # Printing the first few rows for demonstration
                if csv_rows:
                    print("\nFirst 5 rows:")
                    for i, row in enumerate(csv_rows[:5]):
                        print(row)
                    if len(csv_rows) > 5:
                        print("...")
                else:
                    print("The CSV file is empty.")
        else:
            print(f"Error: The CSV file '{csv_filename_in_zip}' was not found inside '{zip_file_path}'.")
            print(f"Please choose one of the files listed above.")

except zipfile.BadZipFile:
    print(f"Error: '{zip_file_path}' is not a valid zip file.")
except FileNotFoundError:
    print(f"Error: The file '{zip_file_path}' was not found. Please ensure it is uploaded or the path is correct.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")


Files found inside 'archive.zip': ['heart_cleveland_upload.csv']

Successfully read 298 rows from 'heart_cleveland_upload.csv'.

First 5 rows:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'condition']
['69', '1', '0', '160', '234', '1', '2', '131', '0', '0.1', '1', '1', '0', '0']
['69', '0', '0', '140', '239', '0', '0', '151', '0', '1.8', '0', '2', '0', '0']
['66', '0', '0', '150', '226', '0', '0', '114', '0', '2.6', '2', '0', '0', '0']
['65', '1', '0', '138', '282', '1', '2', '174', '0', '1.4', '1', '1', '0', '1']
...


In [19]:
import csv
import numpy as np
import zipfile
import pandas as pd

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

#opening the csv from zipfile

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:

#conversion of the csv to table

          data_table = pd.read_csv(file)
          #print(data_table)


#describing the data

print(data_table.columns.to_list(),'\n')
print(data_table.head(),'\n')
print(data_table.info(),'\n')
print(data_table.describe(),'\n')



['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'condition'] 

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   69    1   0       160   234    1        2      131      0      0.1      1   
1   69    0   0       140   239    0        0      151      0      1.8      0   
2   66    0   0       150   226    0        0      114      0      2.6      2   
3   65    1   0       138   282    1        2      174      0      1.4      1   
4   64    1   0       110   211    0        2      144      1      1.8      1   

   ca  thal  condition  
0   1     0          0  
1   2     0          0  
2   0     0          0  
3   1     0          1  
4   0     0          0   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 297 entries, 0 to 296
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   age        297 non-null    int64  
 1   sex

**Inference**



*   The data contains 297 entries with 14 columns
*   The mean age is 54.542088,
           - which mean most of the recorded patients ages are clustered around 54

*   In the 14 columns, only one column contains float values while the other 13 contains integer values
*   Here the mean cholesterol level is around 247.350168;
         - Cholesterol level above 200 suggests that the heart health is in danger. Thus here the cholesterol level is higher.


*   Here the maximum heart rate achieved(thalach) is 149.599327;
         -The maximum heart rate(MHR) is calculated by subtracting the age from 220.
         - Here the mean age is 54.542088;
              - subtracting 54.542088 from 220 results in 165.457912, which is above the MHR of patients
                  -suggesting the patients heart rate is lower than needed




In [30]:
import numpy as np
import pandas as pd
import zipfile

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:
        data_table = pd.read_csv(file)

data_matrix = np.array(data_table)
print(data_matrix)


print(data_table.shape)

[[69.  1.  0. ...  1.  0.  0.]
 [69.  0.  0. ...  2.  0.  0.]
 [66.  0.  0. ...  0.  0.  0.]
 ...
 [35.  1.  3. ...  0.  2.  1.]
 [35.  0.  3. ...  0.  0.  0.]
 [35.  1.  3. ...  0.  2.  1.]]
(297, 14)


**matrix operations**

*   Here the matrix shape is 297,14; that means 297 rows and 14 columns

         - that is since it is not a square matrix,

               - we cannot find its;

                      -Determinant
                      -Inverse
                      -Eigen values and Eigen vectors
                      
              - But we can find it's;
                     -Rank
                     




In [31]:

rank = np.linalg.matrix_rank(data_matrix)
print(rank)

14


**Inference from Rank**



*   Here the Rank is 14;
         -columns are also 14;
                - which means there are 14 linearly independent columns are there and no column(here in this case data categories cannot be reduced)
                       - Thus all columns add meaning to the data
=



**Let's use a SINGLE FUNCTION**



*   From the previous project I have created a single wrapped function 'analyze_matrix' for analysing the key properties of matrix
   - Here we are using that single function

-check out the function here:


****

In [33]:
from matrix_module import analyze_matrix

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:

        analyze_matrix(file)



     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     69    1   0       160   234    1        2      131      0      0.1   
1     69    0   0       140   239    0        0      151      0      1.8   
2     66    0   0       150   226    0        0      114      0      2.6   
3     65    1   0       138   282    1        2      174      0      1.4   
4     64    1   0       110   211    0        2      144      1      1.8   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
292   40    1   3       152   223    0        0      181      0      0.0   
293   39    1   3       118   219    0        0      140      0      1.2   
294   35    1   3       120   198    0        0      130      1      1.6   
295   35    0   3       138   183    0        0      182      0      1.4   
296   35    1   3       126   282    0        2      156      1      0.0   

     slope  ca  thal  condition  
0        1   1     0          0  
1        0   2     